### ISSUE TITLE
How to trigger a deep link from a local notification?

### ISSUE THREAD

**Issue by shimatai** [evollu/react-native-fcm#485]
Hi,

How can I trigger a deep link from a `FCM.presentLocalNotification`?

I'm using React Native 0.45.0 with react-native-fcm 7.0.3 and I'm running in a real Android device and in an emulator too.

All notifications are working well, but I didn't find a way to load a deep link when the user press on the local notification (showed on the Android notification bar).

**Comment by justuseapen** [comment]
I'm working on this now. You can use the notification listener to grab notification details and use your router (in my case react-native-router-flux) to trigger the desired action and show the right view.

If you're using URL schemes seems like Crossroads is the routing lib of choice for projects like this.

**Comment by HAMPLENI** [comment]
@justuseapen 

I was researching whether I would be able to use the notification listener to navigate to a location in the app when the notification is pressed.  From my testing, it looked like I could use the notification listener to do it in all cases except when the app is closed when the user presses the notification.  In that case, it did not hit the notification listener's handler function.  I have the listener being added in the componentDidMount for the component registered into the AppRegistry.  I am also using the Android "custom" remote notifications that actually create local notifications.  This not being able to work is likely a deal breaker for this approach for our project.

Based on this, I am thinking I will need to use click_action.  My teammate offered a clever potential solution of making the click_action native activities just save the information needed about the pressed notification to local storage.  Then, when the app loads, we could just pull from local storage and know what to do in the JavaScript layer.

Your comment makes me wonder if I am missing something about the notification listener approach.

**Comment by shimatai** [comment]
The notification listener doesn't works as I expect, because I want to route to a screen only if the user clicks on the message in the notification bar.

Actually, when app receives a push notification and user clicks on notification, nothing happens.

I know that notification listener can route to a screen, but I want to route only if the user clicks on Android notification bar.

**Comment by evollu** [comment]
@shimatai you can check if the notif has `opened_from_tray`

**Comment by shimatai** [comment]
@evollu 

If the user receives a push notification with the app closed, the registered listener is not being triggered.

What's the solution for this issue? Is there anyway to trigger a method in my app when receiving a push notification with the app closed?

**Comment by evollu** [comment]
put the listener outside of the component lifecycle, just like what readme does.

**Comment by shekharskamble** [comment]
@shimatai - did you manage to resolve this? if yes then how?

**Comment by shimatai** [comment]
@shekharskamble I had to create a Java class that extends `com.evollu.react.fcm.MessagingService` and then you have to override onMessageReceived method.

For example:

`public class MyAppMessagingService extends MessagingService {
    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {
        Set<String> keys = remoteMessage.getData().keySet();
        for (String key : keys) { // List all keys inside incoming message
            Log.i("MyAppMessagingService", "Key " + key + ": " + remoteMessage.getData().get(key));
        }
        super.onMessageReceived(remoteMessage);
    }
}`

**Comment by shekharskamble** [comment]
@shimatai - thanks for the example but how are you triggering deep link from onMessageReceived?

**Comment by shimatai** [comment]
@shekharskamble No, that piece of code is in the native side (it's a Java class that extends the MessagingService of react-native-fcm).

This code is to receive a push notification when the app is closed (not in background nor in use). In my case, when the app receives a notification, my app saves the content in a key (that is read by React Native's AsyncStorage when the app is launched again).

I had to do this because my users can receive notifications while not using the app and the listener written in JavaScript is not triggered.
